docs/articles/tips-and-tricks/backup-restore.rst

```diff
@@ -0,0 +1,185 @@
+called by pytest-mh can be used to implement automatic host backup and restore
+functionality. This is supported out of the box with
+:class:`~pytest_mh.MultihostBackupHost` and
+:class:`~pytest_mh.BackupTopologyController`.
+
+Implementing automatic backup of a host
+=======================================
+
+:class:`~pytest_mh.MultihostBackupHost` is an abstract class that declares
+several abstract methods that have to be implemented:
+
+.. container:: wy-table-responsive
+
+    .. list-table::
+        :widths: 30 70
+        :header-rows: 1
+
+        * - Abstract method name
+          - Description
+        * - :meth:`~pytest_mh.MultihostBackupHost.start`
+          - Start required host services. If no services are needed, this can be
+            implemented as a "no operation" or raise ``NotImplementedError``
+            which will be ignored by internal calls to this method.
+        * - :meth:`~pytest_mh.MultihostBackupHost.stop`
+          - Stop required host services. If no services are needed, this can be
+            implemented as a "no operation" or raise ``NotImplementedError``
+            which will be ignored by internal calls to this method.
+        * - :meth:`~pytest_mh.MultihostBackupHost.backup`
+          - Take backup of the host. The backup can be returned as any Python
+            data, :class:`~pathlib.PurePath` or a sequence of
+            :class:`~pathlib.PurePath`. If the path is returned, it is
+            automatically deleted from the host when all tests are run. If
+            non-path data is returned, any clean up is left on the user if
+            needed -- it is possible to override
+            :meth:`~pytest_mh.MultihostBackupHost.remove_backup`.
+        * - :meth:`~pytest_mh.MultihostBackupHost.restore`
+          - Restore the host from the backup.
+
+The backup is taken automatically during pytest setup and the host is restored
+to this state after each test run. Sometimes, it is not desirable to restore the
+host automatically at this point (for example if this is done by the topology
+controller) and this can be disabled by passing ``auto_restore=False`` to the
+constructor.
+
+.. code-block:: python
+    :caption: Example use of MultihostBackupHost
+
+    class ExampleBackupHost(MultihostBackupHost[MyProjectMultihostDomain]):
+        def __init__(self, *args, **kwargs) -> None:
+            # restore is handled in topology controllers
+            super().__init__(*args, auto_restore=False, **kwargs)
+
+            self.svc: SystemdServices = SystemdServices(self)
+
+    def start(self) -> None:
+        self.svc.start("my-project")
+
+    def stop(self) -> None:
+        self.svc.stop("my-project")
+
+    def backup(self) -> Any:
+        self.logger.info("Creating backup of my-project service")
+
+        # yields backup path
+        result = self.conn.run("my-project create-backup", log_level=ProcessLogLevel.Error)
```

Comment:
From the documentation I was expecting to see a path at this point (i.e. /etc/pam.d/). Did I misunderstand the documentation, or is the example somewhat confusing?

Reply:
Not sure, you tell me :-)

You don't have to necessarily just backup paths, but you can run commands (e.g. ipa-backup, ipa-restore) or just set local data (ldap).